agents/magpie/magpie_agent.py

```diff
@@ -0,0 +1,705 @@
+                # not block
+                self.out_queue.put(f)
+        return True, "Stopped read process"
+
+    def _stop_read(self, session, params=None):
+        self._running = False
+        return True, "Stopping read process"
+
+
+    def stream_fake_data(self, session, params=None):
+        """stream_fake_data()
+
+        **Process** - Process for streaming fake data. This will queue up
+        G3Frames full of fake data to be sent to lyrebird.
+        """
+        self._run_fake_stream = True
+        ndets = self.fp.num_dets
+        chans = np.arange(ndets)
+        frame_start = time.time()
+        while self._run_fake_stream:
+            time.sleep(2)
+            frame_stop = time.time()
+            ts = np.arange(frame_start, frame_stop, 1./self.target_rate)
+            frame_start = frame_stop
+            nframes = len(ts)
+
+            data_out = np.random.normal(0, 1, (nframes, ndets))
+            data_out += np.sin(2*np.pi*ts[:, None] + .2 * chans[None, :])
+
+            for t, d in zip(ts, data_out):
+                fr = core.G3Frame(core.G3FrameType.Scan)
+                fr['idx'] = 0
+                fr['data'] = core.G3VectorDouble(d)
+                fr['timestamp'] = core.G3Time(t * core.G3Units.s)
+                self.out_queue.put(fr)
+
+                fr = core.G3Frame(core.G3FrameType.Scan)
+                fr['idx'] = 1
+                fr['data'] = core.G3VectorDouble(np.sin(d))
+                fr['timestamp'] = core.G3Time(t * core.G3Units.s)
+                self.out_queue.put(fr)
+
+        return True, "Stopped fake stream process"
+
+
+    def _stop_stream_fake_data(self, session, params=None):
+        self._run_fake_stream = False
+        return True, "Stopping fake stream process"
+
+
+    @ocs_agent.param('dest', type=int)
+    def send(self, session, params=None):
+        """
+        **Process** - Process for sending outgoing G3Frames. This will query
+        the out_queue for frames to be sent to lyrebird. This will try to
+        regulate how fast it sends frames such that the delay between when the
+        frames are sent, and the timestamp of the frames are fixed.
+
+        """
+        self._send_running = True
+        session.set_status('running')
+
+        first_frame_time = None
+        stream_start_time = None
+
+        sender = core.G3NetworkSender(
+            hostname='*', port=params['dest'], max_queue_size=1000
+        )
+
+        sender.Process(self.fp.config_frame())
+        while self._send_running:
+            f = self.out_queue.get(block=True)
+            t = f['timestamp'].time / core.G3Units.s
+            now = time.time()
+            if first_frame_time is None:
+                first_frame_time = t
+                stream_start_time = now
+
+            this_frame_time = stream_start_time + (t - first_frame_time) + self.delay
+            if this_frame_time > now:
+                time.sleep(this_frame_time - now)
```

Comment:
Is the `session` object the same in the process start / stop function? I think maybe instead of having like:
```
while self._send_running:
    time.sleep(...)
``` 
we should do have something like:
```
while session.status == 'running':
    ocs_agent.sleep_until_stopped(10000, session)
    ...
```
and then process stop functions should just set `session.status = 'stopping'`

Reply:
And more generally, a "default" stop function that moves status to 'stopping' also makes sense.

I think we'd discussed also having status advance to "running" automatically before calling the start function, so people don't have to call it themselves.  I can't find the discussion on that anywhere, but it was the plan IIRC.

Reply:
> This seems to come up in a bunch of different contexts (especially for dealing with smurf stuff and streaming data). Is there any way to come up with a generic way to do this? Like a function that sleeps until either a certain duration has passed or the current operation is stopped?

You'd need to "break" manually ... main loops would look like:
```
  while True:
    if not ocs_agent.sleep_while_running(10000, session):
      break
    ...
```
but ya, I think that's useful.  Get it working here and we can migrate it into ocs.